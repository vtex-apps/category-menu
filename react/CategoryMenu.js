import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { NoSSR } from 'render'
import { graphql } from 'react-apollo'

import getCategories from './queries/categoriesQuery.gql'
import CategoryItem from './components/CategoryItem'
import LoadingBar from './components/LoadingBar'
import { categoryPropType } from './propTypes'

import './global.css'

/**
 * Component that represents the menu containing the categories of the store
 */
export class CategoryMenu extends Component {
  static propTypes = {
    /** Whether to show the promotion category or not */
    showPromotionCategory: PropTypes.bool,
    /** Whether to show the gift category or not */
    showGiftCategory: PropTypes.bool,
    data: PropTypes.shape({
      loading: PropTypes.bool.isRequired,
      categories: PropTypes.arrayOf(categoryPropType),
    }),
  }

  static defaultProps = {
    showPromotionCategory: false,
    showGiftCategory: false,
  }

  static schema = {
    title: 'Category Menu',
    description:
      'A menu showing a list of the available categories on the store',
    type: 'object',
    properties: {
      showPromotionCategory: {
        type: 'boolean',
        title: 'Show the promotion category',
      },
      showGiftCategory: {
        type: 'boolean',
        title: 'Show the gifts category',
      },
    },
  }

  render() {
    const {
      data: { categories, loading },
    } = this.props

    const fallback = (
      <div className="vtex-category-menu">
        <nav className="flex w-two-thirds center h-100">
          <CategoryItem category={{ id: 0, name: 'Category', href: '#category' }} />
        </nav>
      </div>
    )

    return (
      <NoSSR onSSR={fallback}>
        <div className="vtex-category-menu">
          <LoadingBar loading={loading}>
            <nav className="flex w-two-thirds center h-100">
              {categories &&
                categories.map(category => (
                  <CategoryItem key={category.id} category={category} />
                ))}
            </nav>
          </LoadingBar>
        </div>
      </NoSSR>
    )
  }
}

export default graphql(getCategories)(CategoryMenu)
