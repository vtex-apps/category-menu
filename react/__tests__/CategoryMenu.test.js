import React from 'react'
import { mount } from 'enzyme'

import { CategoryMenu } from '../CategoryMenu'

describe('CategoryMenu component', () => {
  let wrapper

  beforeEach(() => {
    const mockedCategories = [
      {
        id: 1,
        href: '#1',
        name: 'Category 1',
        slug: 'category-1',
        hasChildren: false,
        children: [],
      },
      {
        id: 2,
        href: '#2',
        name: 'Category 2',
        slug: 'category-2',
        hasChildren: false,
        children: [],
      },
      {
        id: 3,
        href: '#3',
        name: 'Category 3',
        slug: 'category-3',
        hasChildren: false,
        children: [],
      },
    ]

    wrapper = mount(
      <CategoryMenu
        data={{
          categories: mockedCategories,
          loading: false,
        }}
      />
    )
  })

  it('should be rendered', () => {
    expect(wrapper).toBeDefined()
  })

  it('should match snapshot', () => {
    expect(wrapper).toMatchSnapshot()
  })

  it('should render 4 menu items', () => {
    expect(wrapper.find('.vtex-category-menu__item').length).toBe(4)
  })

  it('should have all provided categories', () => {
    expect(
      wrapper.containsAllMatchingElements([
        /* eslint-disable react/jsx-key */
        <a>Category 1</a>,
        <a>Category 2</a>,
        <a>Category 3</a>,
        /* eslint-enable */
      ])
    ).toBe(true)
  })

  it("shouldn't be able to find a `Category 4` item", () => {
    expect(wrapper.containsMatchingElement(<a>Category 4</a>)).toBe(false)
  })
})
