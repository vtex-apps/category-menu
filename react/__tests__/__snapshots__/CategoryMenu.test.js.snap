// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CategoryMenu component should match snapshot 1`] = `
<CategoryMenu
  data={
    Object {
      "categories": Array [
        Object {
          "children": Array [],
          "hasChildren": false,
          "href": "#1",
          "id": 1,
          "name": "Category 1",
          "slug": "category-1",
        },
        Object {
          "children": Array [],
          "hasChildren": false,
          "href": "#2",
          "id": 2,
          "name": "Category 2",
          "slug": "category-2",
        },
        Object {
          "children": Array [],
          "hasChildren": false,
          "href": "#3",
          "id": 3,
          "name": "Category 3",
          "slug": "category-3",
        },
      ],
      "loading": false,
    }
  }
  showGiftCategory={false}
  showPromotionCategory={false}
>
  <NoSSR
    onSSR={
      <div
        className="vtex-category-menu"
      >
        <nav
          className="flex w-two-thirds center h-100"
        >
          <CategoryItem
            category={
              Object {
                "href": "#category",
                "id": 0,
                "name": "Category",
              }
            }
          />
        </nav>
      </div>
    }
  >
    <div
      className="vtex-category-menu"
    >
      <LoadingBar
        fallback={null}
        loading={false}
      >
        <nav
          className="flex w-two-thirds center h-100"
        >
          <CategoryItem
            category={
              Object {
                "children": Array [],
                "hasChildren": false,
                "href": "#1",
                "id": 1,
                "name": "Category 1",
                "slug": "category-1",
              }
            }
            key="1"
          >
            <div
              className="vtex-category-menu__item ph4"
              data-testid="category-item"
            >
              <a
                className="db no-underline ttu"
                href="#1"
              >
                Category 1
              </a>
            </div>
          </CategoryItem>
          <CategoryItem
            category={
              Object {
                "children": Array [],
                "hasChildren": false,
                "href": "#2",
                "id": 2,
                "name": "Category 2",
                "slug": "category-2",
              }
            }
            key="2"
          >
            <div
              className="vtex-category-menu__item ph4"
              data-testid="category-item"
            >
              <a
                className="db no-underline ttu"
                href="#2"
              >
                Category 2
              </a>
            </div>
          </CategoryItem>
          <CategoryItem
            category={
              Object {
                "children": Array [],
                "hasChildren": false,
                "href": "#3",
                "id": 3,
                "name": "Category 3",
                "slug": "category-3",
              }
            }
            key="3"
          >
            <div
              className="vtex-category-menu__item ph4"
              data-testid="category-item"
            >
              <a
                className="db no-underline ttu"
                href="#3"
              >
                Category 3
              </a>
            </div>
          </CategoryItem>
        </nav>
      </LoadingBar>
    </div>
  </NoSSR>
</CategoryMenu>
`;
